Dalamud/DiscordBot/DiscordBotManager.cs

```diff
@@ -315,12 +348,187 @@ public async Task ProcessCfPreferredRoleChange(string rouletteName, string prevR
+                {"\uE0bd", "♋️" }, // level sync icon reversed
+                {"\uE0be", "🔽" }, // sync down icon
+                {"\uE0bf", "☒" }, // x in diamond
+
+                {"\uE0c0", "🌟" }, // wide star
+                {"\uE0c1", "Ⅰ" }, // wide roman numeral I
+                {"\uE0c2", "Ⅱ" }, // wide roman numeral II
+                {"\uE0c3", "Ⅲ" }, // wide roman numeral III
+                {"\uE0c4", "Ⅳ" }, // wide roman numeral IV
+                {"\uE0c5", "Ⅴ" }, // wide roman numeral V
+                {"\uE0c6", "Ⅵ" }, // wide roman numeral VI
+
+                {"\uE0D0", "Lᴛ" }, //LocalTimeEn   
+                {"\uE0D1", "Sᴛ" }, //ServerTimeEn   
+                {"\uE0D2", "Eᴛ" }, //EorzeaTimeEn   
+                {"\uE0D3", "Oᴢ" }, //LocalTimeDe   
+                {"\uE0D4", "Sᴢ" }, //ServerTimeDe   
+                {"\uE0D5", "Eᴢ" }, //EorzeaTimeDe   
+                {"\uE0D6", "Hʟ" }, //LocalTimeFr   
+                {"\uE0D7", "Hs" }, //ServerTimeFr   
+                {"\uE0D8", "Hᴇ" }, //EorzeaTimeFr   
+                {"\uE0D9", "本" }, //LocalTimeJa   
+                {"\uE0DA", "服" }, //ServerTimeJa   
+                {"\uE0DB", "艾" }, //EorzeaTimeJa   
+            };
+            // Transform the block letters into regional indicator letters. Will combine with flags if they are together in the right combo, so also adding a zero-width space
+            for( var i = 0; i < 26; i++ )
+            {
+                char xivchar = (char)(0xe071 + i);
+                string unichar = char.ConvertFromUtf32('A' + 0x1f1a5 + i);
+                string zerowidth = "\u200B";
+
+                ffxivSpecialCharacters.Add(xivchar.ToString(), unichar + zerowidth);
+            }
+            
+            // Number squares changed to enclosed number unicode. ⓪ is special, then ①-⑳ are one sequence then ㉑-㊿ are yet another, however xiv only goes to 31
+            for ( var i = 0; i <= 31; i++ )
+            {
+                char xivchar = (char)(0xe08f + i);
+                string unichar = "";
+                if( i == 0 )
+                {
+                    unichar = char.ConvertFromUtf32('⓪' + i);
+                }
+                if( i > 0 && i <= 19 )
+                {
+                    unichar = char.ConvertFromUtf32('①' + i - 1);
+                }
+                if( i >= 21 )
+                {
+                    unichar = char.ConvertFromUtf32('㉑' + i - 21);
+                }
+
+                ffxivSpecialCharacters.Add(xivchar.ToString(), unichar);
+            }
+
+            string returnString = "";
+            foreach (char c in ffxivString)
+            {
+                if ( ffxivSpecialCharacters.ContainsKey(c.ToString()) )
+                {
+                    returnString += ffxivSpecialCharacters[c.ToString()];
+                } else
+                {
+                    returnString += c;
                 }
             }
```

Comment:
Please do this with a string builder.  This is already a pretty inefficient way to modify a string, but there's no reason to not at least use a builder to be nice to the system.

Reply:
A string builder would be extremely inefficient. At this level of text processing there is literally nothing faster than `O(n)` for a string with `O(1)` for each lookup, which is what this is. I initially did it with a builder and a series of replaces when it was just a few characters in an experiment, but it quickly became something that wouldn't look out of place in Yandere Simulator.

Reply:
The purpose of the builder is to not have to copy the string potentially many many times, and to avoid as much memory fragmentation.  It is absolutely going to be more efficient than a direct append

Reply:
This is the part where it looks at the string that is being immediately processed, as in it is preparing a chat message to send to the discord bot, and will not be doing it "many many times". I am working on a change to only build the lookup dictionary once, but for this part we are processing the incoming message for output and there is no more efficient way to do that.

Reply:
Every _single_ character append to a string has the potential to result in a copy of the entire string.  This is one of the major reasons string builders _exist_ in many languages.

In reality it's going to be affected by alignment and padding chosen by (in this case) the managed runtime, but it's absolutely going to happen.

I literally just mean to replace your string += with a stringbuilder Append(), and don't call ToString() until the return.  It's not a rework, and it is unequivocally more efficient.

Reply:
what

do you understand the scope of this part of the code

this is literally the part where someone says "i like :hqsymbol: butts" in fc chat, then it looks through the message to find the ffxiv special characters, in this case, swapping out the hq symbol as necessary

Reply:
Yes I understand exactly what it does.

I'm not sure you understand what buliders are for.  If you are this confused, just let this get merged and I'll fix it.

Reply:
There's also the issue with formatting being weird making this PR a bit hard to see actual changes, but that's probably okay and I know some IDEs probably automatically do it

Reply:
string + string isn't O(1) btw, its O(n^2)

Reply:
oh so c# concatenates strings in a dumb way

alright again this is... a really really stupid thing to nitpick over as none of this is anywhere close to the volume or processing power to where this much energy and time needs to be spent obsessing over it, but nevertheless i have an idea for how to fix it

Reply:
I don't believe there exists any language that string + string is O(1)

Reply:
I mean.. every language with string concat works that way.

I think maybe you think I meant changing more than I'm actually suggesting.  The flow there is fine, it's just this:
`var returnString = new StringBuilder();
foreach (char c in ffxivString)
{
    if ( ffxivSpecialCharacters.ContainsKey(c.ToString()) )
    {
        returnString.Append(ffxivSpecialCharacters[c.ToString()]);
    }
    else
    {
        returnString.Append(c);
    }
}

return returnString.ToString();
`

Reply:
Like you *could* rework things here to be greedy about in-line changes, but since the replacements are variable-sized, that really doesn't seem worth it (or clean).  I mentioned the builder here because this is literally the stereotypical use case they are for, and it's a simple fix.

Reply:
Updated. It now creates a stringbuilder with a copy of the message, then does the loop to look for things that need replacing, then runs a stringbuilder replace when those items are found.